external-crates/move/crates/move-regex-borrow-graph/src/collections.rs

```diff
@@ -0,0 +1,510 @@
+                for s in self
+                    .node(imm_new_ref)
+                    .unwrap()
+                    .successors()
+                    .filter(|s| s != imm_new_ref)
+                {
+                    // s is new ==> s is imm
+                    debug_assert!(!all_new_refs.contains(&s) || imm_new_refs.contains(&s));
+                    debug_assert!(!mut_new_refs.contains(&s));
+                }
+            }
+        }
+        Ok(new_refs)
+    }
+
+    fn extend_by_extension(
+        &mut self,
+        loc: Loc,
+        sources: impl IntoIterator<Item = Ref>,
+        ext: Extension<Lbl>,
+        new_ref: Ref,
+        exclude: &BTreeSet<Ref>,
+    ) -> Result<Ref> {
+        self.check_invariant();
+        let mut edges_to_add = vec![];
+        for x in sources {
+            debug_assert!(!exclude.contains(&x));
+            self.determine_new_edges(&mut edges_to_add, x, &ext, new_ref, exclude)?;
+        }
+        for (p, r, s) in edges_to_add {
+            debug_assert!(p == new_ref || s == new_ref);
+            self.add_edge(p, loc, r, s)?;
+        }
+        self.check_invariant();
+        Ok(new_ref)
+    }
+
+    fn determine_new_edges(
+        &self,
+        edges_to_add: &mut Vec<(Ref, Regex<Lbl>, Ref)>,
+        x: Ref,
+        ext: &Extension<Lbl>,
+        new_ref: Ref,
+        exclude: &BTreeSet<Ref>,
+    ) -> Result<()> {
+        for y in self
+            .node(&x)?
+            .predecessors()
+            .filter(|y| !exclude.contains(y))
+        {
+            for y_to_x in self.node(&y)?.regexes(&x)? {
+                edges_to_add.push((y, y_to_x.clone().extend(ext), new_ref))
+            }
+        }
+        for y in self.node(&x)?.successors().filter(|y| !exclude.contains(y)) {
+            for x_to_y in self.node(&x)?.regexes(&y)? {
+                // For the edge x --> y, we adding a new edge x --> new_ref
+                // In cases of a label extension, we might need to add an edge new_ref --> y
+                // if the extension is a prefix of x_to_y.
+                // However! In cases where an epsilon or dot-star is involved,
+                // we might also have the case that we can remove x --> y as a prefix of
+                // x --> new_ref
+                // In the case where we have `e.remove_prefix(p)` and `e` is a list of labels
+                // `fgh` and `p` is `.*`, we will consider all possible suffixes of `e`,
+                // `[fgh, gh, h, epsilon]`. This could grow rather quickly, so we might
+                // want to optimize this representation
+                for x_to_y_suffix in x_to_y.remove_prefix(ext) {
+                    edges_to_add.push((new_ref, x_to_y_suffix, y))
+                }
+                for regex_suffix in ext.remove_prefix(x_to_y) {
+                    edges_to_add.push((y, regex_suffix, new_ref));
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn add_edge(
```

Comment:
Nit: call these args `source` and `target` instead of `predecessor` and `successor`.